File: jarvis-desktop/src/memory-used.js
/**
 * Temporary chat, "Used in this answer" and "Jarvis remembered N things" -
 * the owner's decisions of 2026-09-25 (JARVIS-API.md sections 4, 6, 18.1
 * and 19.5).
 *
 * TEMPORARY CHAT. One tap in the quickbar. While it is on, every question
 * is sent with `temporary: true`, and the PC recalls no facts (no pinned
 * list either), learns nothing (no "Remember:" either) and keeps nothing in
 * the chat history. Tools, approval cards and the rest are unchanged, so it
 * needs no card: it only makes things stricter. Turning it on or off starts
 * a new conversation, so nothing said in one kind is re-sent in the other.
 * A PC without it says so before anything is sent (commands.rs
 * `temporary_chat_available`, and `stream_chat` checks again); an answer
 * whose X-Jarvis-Route does not say `temporary: true` gets a plain warning
 * rather than a pretence.
 *
 * USED IN THIS ANSWER. X-Jarvis-Route lists the facts an answer used by id
 * only (commands.rs passes them on as `memory_ids`, numbers). A quiet line
 * under the answer, "Used 2 memories", opens the list: the words are read
 * from the PC by id then (brain/used.rs `memory_used` - hidden, like every
 * memory list, while Windows Hello hides them), each with Forget and "Erase
 * the words" (the Brain's one-fact commands, held on a stale link, asked
 * about first in the words both apps use). A pinned fact says so.
 *
 * The Brain's "Jarvis remembered 2 things" line opens the same list for the
 * facts automatic learning just saved (the `memory_saved` event's ids).
 *
 * This module holds the words and reads what the PC sends; main.js and
 * brain.js draw. The phone says the same words (net/MemoryUsed.kt).
 *
 * @module memory-used
 */

/** The mode's name, and the one line on the empty chat - both apps. */
export const TEMPORARY_LABEL = "Temporary chat";
export const TEMPORARY_LINE =
  "Temporary chat: Jarvis won't use or learn from your memory, and this chat isn't kept.";
/** The toggle's two titles. */
export const TEMPORARY_ON_TITLE =
  "Start a temporary chat: Jarvis won't use or learn from your memory, and it isn't kept";
export const TEMPORARY_OFF_TITLE = "End the temporary chat and start a normal one";
/** Said when the owner turns it on or off (a new conversation each time). */
export const TEMPORARY_STARTED = "Temporary chat started. Nothing in it is remembered or kept.";
export const TEMPORARY_ENDED = "Temporary chat ended. Your next question starts a normal chat.";

/** A PC whose backend has no temporary chat - commands.rs
 *  TEMPORARY_UNAVAILABLE, word for word. Nothing is sent. */
export const TEMPORARY_UNAVAILABLE =
  "Temporary chat isn't available on this PC's version of Jarvis, so nothing was sent. " +
  "Run apply-patches.ps1 on the PC to update it.";

/** An answer to a temporary question whose route header does not confirm
 *  it - never pretend it was. */
export const TEMPORARY_NOT_CONFIRMED =
  "Your PC did not confirm this was a temporary chat, so this answer may have used your " +
  "memory and the chat may be kept.";

/** A "Remember: ..." in a temporary chat (the route's `remember_off`). */
export const REMEMBER_OFF = "Remember: is off in a temporary chat.";

/**
 * What the route header says about a temporary question, once it is in:
 * "confirmed" (`temporary: true`), "unconfirmed" (anything else - an older
 * PC, or memory the PC cannot switch off), or "" for a normal question.
 */
export function temporaryOutcome(sentTemporary, route) {
  if (!sentTemporary) return "";
  return route && typeof route === "object" && route.temporary === true
    ? "confirmed"
    : "unconfirmed";
}

/** The facts an answer used, as the route line carries them (`memory_ids`,
 *  whole numbers above 0), each once. */
export function usedIds(route) {
  const ids = route && Array.isArray(route.memory_ids) ? route.memory_ids : [];
  const out = [];
  for (const id of ids) if (Number.isInteger(id) && id > 0 && !out.includes(id)) out.push(id);
  return out;
}

/** "Used 1 memory" / "Used 2 memories" - the quiet line under an answer. */
export function usedLine(n) {
  const count = Number.isInteger(n) && n > 0 ? n : 0;
  if (!count) return "";
  return `Used ${count} ${count === 1 ? "memory" : "memories"}`;
}

/** The list's titles: under an answer, and in the Brain. */
export const USED_TITLE = "Used in this answer";
export const REMEMBERED_TITLE = "Remembered just now";
export const USED_LINE_TITLE = "Show which of your saved facts this answer used.";
export const REMEMBERED_LINE_TITLE = "Show what Jarvis just saved automatically.";
/** The Brain's way on to the whole list, as before. */
export const SHOW_ALL = "Show everything saved automatically";

/** What each fact can say beside its words. */
export const PINNED_MARK = "pinned";
export const PINNED_MARK_TITLE = "On \"Always keep in mind\": Jarvis reads it with every question.";
export const NOT_CURRENT_MARK = "no longer in use";
export const NOT_CURRENT_TITLE =
  "Jarvis no longer uses this fact (it was forgotten, corrected or ran out). An answer about " +
  "the past can still mention it.";

/** A PC without the read route (Rust answers `{available: false}`). */
export const USED_MISSING =
  "Your PC's Jarvis cannot show which facts these were yet - run apply-patches.ps1 on the PC " +
  "to update it.";

/** While Windows Hello hides the memory lists. The quickbar cannot ask
 *  Windows Hello itself: Show is in the Brain. */
export function hiddenLine(n, where = "quickbar") {
  const count = Number.isInteger(n) && n > 0 ? n : 0;
  const what = count === 1 ? "This fact is" : "These facts are";
  const how = where === "brain"
    ? "hidden until Windows Hello confirms it is you."
    : "hidden until Windows Hello confirms it is you - press Show on the Brain's Memory tab.";
  return `${what} ${how}`;
}

/** Ids the PC had no fact for at all. */
export function missingLine(n) {
  const count = Number.isInteger(n) && n > 0 ? n : 0;
  if (!count) return "";
  return count === 1
    ? "1 of them is no longer on this PC."
    : `${count} of them are no longer on this PC.`;
}

const num = (v) => (typeof v === "number" && Number.isFinite(v) ? v : null);
const text = (v) => (typeof v === "string" ? v : "");

/**
 * `memory_used`'s answer, read.
 *
 * `{facts: [{id, text, current, pinned, created, valid_to, erased_at}],
 * missing: [id]}` from the PC; `available: false` (with `why`) from an
 * older one; `hidden: true` with `hidden_count` while Windows Hello hides
 * the memory lists (Rust took the facts out). A fact without a whole-number
 * id is dropped, never guessed. An erased fact never has words here: the
 * PC sends none, and even a stray "[erased]" is not shown.
 */
export function readUsed(answer) {
  const a = answer && typeof answer === "object" ? answer : {};
  const available = a.available !== false && Array.isArray(a.facts);
  const facts = (available ? a.facts : [])
    .filter((f) => f && Number.isInteger(f.id) && f.id > 0)
    .map((f) => {
      const erasedAt = num(f.erased_at);
      return {
        id: f.id,
        text: erasedAt ? "" : text(f.text),
        current: f.current === true && !erasedAt,
        pinned: f.pinned === true && f.current === true && !erasedAt,
        validTo: num(f.valid_to),
        erasedAt,
      };
    });
  const missing = (available && Array.isArray(a.missing) ? a.missing : [])
    .filter((id) => Number.isInteger(id));
  return {
    available,
    why: available ? "" : text(a.why) || USED_MISSING,
    hidden: available && a.hidden === true,
    hiddenCount: num(a.hidden_count) ?? 0,
    facts,
    missing,
  };
}

/**
 * What one row offers: Forget only on a fact still in use (a forgotten one
 * has nothing to forget), and "Erase the words" on any fact that still has
 * its words - the desktop erases forgotten facts too, as in the Brain.
 */
export function rowActions(f) {
  return {
    forget: Boolean(f && f.current && !f.erasedAt),
    erase: Boolean(f && !f.erasedAt),
  };
}

File: jarvis-client/app/src/main/java/com/jarvis/client/net/MemoryUsed.kt
package com.jarvis.client.net

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.longOrNull

/**
 * A temporary chat - the owner's decision of 2026-09-25 (docs/JARVIS-API.md
 * section 18.1).
 *
 * One tap on Home. While it is on, every question - typed or spoken - goes
 * with `"temporary": true`, and the PC recalls no facts (no pinned list
 * either), learns nothing (no "Remember:" either) and keeps nothing in the
 * chat history. Tools, approval cards and the rest are unchanged, so it needs
 * no card: it only makes things stricter. Turning it on or off starts a new
 * conversation ([ChatSession.setTemporary]), so nothing said in one kind of
 * chat is re-sent in the other.
 *
 * Offered only when `/api/version` says the PC has it
 * (`capabilities.temporary_chat`); otherwise the phone says [UNAVAILABLE] and
 * sends nothing as temporary. An answer whose `X-Jarvis-Route` does not say
 * `"temporary": true` gets [NOT_CONFIRMED] rather than a pretence.
 *
 * The desktop says the same words (jarvis-desktop/src/memory-used.js).
 * Pure Kotlin, no Android types, so `MemoryUsedTest` runs it on a plain JVM.
 */
object TemporaryChat {
    /** The capability `/api/version` reports (`jarvis_events._capability_probe`). */
    const val CAPABILITY = "temporary_chat"

    /** The mode's name, and the one line on the empty chat - both apps. */
    const val LABEL = "Temporary chat"
    const val LINE =
        "Temporary chat: Jarvis won't use or learn from your memory, and this chat isn't kept."

    /** The toggle, when off, and what ends it. */
    const val START = "Temporary chat"
    const val END = "End"

    /** Said when the owner turns it on or off (a new conversation each time). */
    const val STARTED = "Temporary chat started. Nothing in it is remembered or kept."
    const val ENDED = "Temporary chat ended. Your next question starts a normal chat."

    /** A PC whose backend has no temporary chat. Nothing is sent. */
    const val UNAVAILABLE =
        "Temporary chat isn't available on this PC's version of Jarvis, so nothing was sent. " +
            "Run apply-patches.ps1 on the PC to update it."

    /** An answer to a temporary question whose route header does not confirm it. */
    const val NOT_CONFIRMED =
        "Your PC did not confirm this was a temporary chat, so this answer may have used your " +
            "memory and the chat may be kept."

    /** A "Remember: ..." in a temporary chat (the route's `remember_off`). */
    const val REMEMBER_OFF = "Remember: is off in a temporary chat."

    /** The `/api/chat` body's flag: present, and true, only for a temporary chat. */
    const val FIELD = "temporary"

    /**
     * The lines to show under an answer, from its `X-Jarvis-Route` header:
     * [NOT_CONFIRMED] when the question went as temporary and the header
     * does not say `"temporary": true`, and [REMEMBER_OFF] when it says
     * `"remember_off": true`. Empty for an ordinary question.
     */
    fun notes(sentTemporary: Boolean, header: String?): List<String> {
        if (!sentTemporary) return emptyList()
        val route = MemoryUsed.route(header)
        return listOfNotNull(
            if (route?.flag("temporary") == true) null else NOT_CONFIRMED,
            if (route?.flag("remember_off") == true) REMEMBER_OFF else null,
        )
    }

    private fun JsonObject.flag(key: String): Boolean? =
        (this[key] as? JsonPrimitive)?.takeIf { it !is JsonNull && !it.isString }?.booleanOrNull
}

/**
 * "Used in this answer", and the facts behind "Jarvis remembered N things" -
 * the owner's decisions of 2026-09-25 (docs/JARVIS-API.md sections 4, 6 and
 * 19.5).
 *
 * `X-Jarvis-Route` lists the facts an answer used by id only
 * (`injected_ids`, "mem:<id>"), and the `memory_saved` event lists what
 * automatic learning saved by id only. When the owner opens "Used 2
 * memories" under an answer, or "Jarvis remembered 2 things" in Mind, their
 * words are read from the PC by id - `GET /api/memory/used?ids=12,15` - and
 * each fact still in use gets a Forget, with the same confirm as Mind's
 * Forget. A pinned fact says so. Hidden, like Mind's other memory lists,
 * under "Hide memory lists and chat history".
 *
 * "Erase the words" is not offered here on the phone: on the phone it stays
 * where it is (Mind -> Saved automatically), and a fact used in an answer may
 * be one the phone does not otherwise list (docs/ARCHITECTURE.md section 8).
 * The desktop offers both.
 */
object MemoryUsed {
    const val PATH = "/api/memory/used"

    /** The PC's own limit on one read (`jarvis_memory.USED_MAX`). */
    const val MAX = 100

    /** The list's titles: under an answer, and in Mind. */
    const val USED_TITLE = "Used in this answer"
    const val REMEMBERED_TITLE = "Remembered just now"

    /** Mind's way on from those facts to the whole list, as before. */
    const val SHOW_ALL = "Show everything saved automatically"

    /** What each fact can say beside its words. */
    const val PINNED_MARK = "pinned"
    const val NOT_CURRENT_MARK = "no longer in use"

    /** A PC without the read route (a 404 or a 501). The desktop says the same. */
    const val MISSING =
        "Your PC's Jarvis cannot show which facts these were yet - run apply-patches.ps1 on the PC " +
            "to update it."

    /** While "Hide memory lists and chat history" hides them. */
    const val HIDDEN = "Hidden. Tap Show and confirm it is you."

    /** On a stale link: Forget waits. */
    const val HELD = "Not connected to the desktop, so Forget waits until the link is back."

    /** "Used 1 memory" / "Used 2 memories" - the quiet line under an answer. Null for none. */
    fun usedLine(count: Int): String? = when {
        count <= 0 -> null
        count == 1 -> "Used 1 memory"
        else -> "Used $count memories"
    }

    /** Ids the PC had no fact for at all. Null for none. */
    fun missingLine(count: Int): String? = when {
        count <= 0 -> null
        count == 1 -> "1 of them is no longer on this PC."
        else -> "$count of them are no longer on this PC."
    }

    /** `X-Jarvis-Route` as a JSON object, or null. */
    internal fun route(header: String?): JsonObject? {
        if (header.isNullOrBlank()) return null
        return runCatching { JarvisJson.parseToJsonElement(header.trim()) as? JsonObject }.getOrNull()
    }

    /**
     * The facts an answer used, from its `X-Jarvis-Route`: each "mem:<id>"
     * in `injected_ids` as a whole number above 0, in order, each once, at
     * most [MAX]. "fact:<n>" (the older word list, which has no id) and
     * anything else are left out.
     */
    fun idsFromRouteHeader(header: String?): List<Long> {
        val ids = route(header)?.get("injected_ids") as? JsonArray ?: return emptyList()
        val out = LinkedHashSet<Long>()
        for (el in ids) {
            val s = (el as? JsonPrimitive)?.takeIf { it.isString }?.content ?: continue
            val digits = s.removePrefix("mem:").takeIf { s.startsWith("mem:") } ?: continue
            if (digits.isEmpty() || digits.length > 12 || !digits.all { it in '0'..'9' }) continue
            val n = digits.toLongOrNull()?.takeIf { it > 0 } ?: continue
            out.add(n)
            if (out.size >= MAX) break
        }
        return out.toList()
    }

    /** The read for these ids, or null when there are none (or too many) to ask for. */
    fun path(ids: List<Long>): String? {
        val list = ids.filter { it > 0 }.distinct()
        if (list.isEmpty() || list.size > MAX) return null
        return PATH + "?ids=" + list.joinToString(",")
    }

    data class Fact(
        val id: Long,
        /** The words; "" for an erased fact, whose words are gone. */
        val text: String,
        val current: Boolean,
        val pinned: Boolean,
        val validTo: Double?,
        val erasedAt: Double?,
    ) {
        /** Forget only on a fact still in use: a forgotten one has nothing to forget. */
        val canForget: Boolean get() = current && erasedAt == null
    }

    data class View(val facts: List<Fact>, val missing: List<Long>)

    /** What one read came to. */
    sealed interface Read {
        data class Shown(val view: View) : Read
        /** A PC without the route. */
        data object Missing : Read
        data class Failed(val why: String) : Read
    }

    /**
     * `GET /api/memory/used`, read - or null when it is not that answer (no
     * `facts` array). A fact without a whole-number id is dropped, never
     * guessed. An erased fact never has words here: the PC sends none, and
     * even a stray "[erased]" is not shown.
     */
    fun parse(body: JsonObject): View? {
        val raw = body["facts"] as? JsonArray ?: return null
        val facts = raw.mapNotNull { el ->
            val o = el as? JsonObject ?: return@mapNotNull null
            val id = o.number("id")?.longOrNull?.takeIf { it > 0 } ?: return@mapNotNull null
            val erasedAt = o.number("erased_at")?.doubleOrNull?.takeIf { it.isFinite() && it > 0 }
            val current = o.flag("current") == true && erasedAt == null
            Fact(
                id = id,
                text = if (erasedAt != null) "" else o.text("text").orEmpty(),
                current = current,
                pinned = current && o.flag("pinned") == true,
                validTo = o.number("valid_to")?.doubleOrNull?.takeIf { it.isFinite() },
                erasedAt = erasedAt,
            )
        }
        val missing = (body["missing"] as? JsonArray).orEmpty().mapNotNull {
            (it as? JsonPrimitive)?.takeIf { p -> p !is JsonNull && !p.isString }?.longOrNull
        }
        return View(facts, missing)
    }

    /** A read that failed because this PC has no such route: a 404 or a 501. */
    fun missing(error: ApiError): Boolean =
        error == ApiError.NotFound || (error is ApiError.Server && error.code == 501)

    private fun JsonObject.number(key: String): JsonPrimitive? =
        (this[key] as? JsonPrimitive)?.takeIf { it !is JsonNull && !it.isString }

    private fun JsonObject.text(key: String): String? =
        (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.contentOrNull

    private fun JsonObject.flag(key: String): Boolean? =
        (this[key] as? JsonPrimitive)?.takeIf { it !is JsonNull && !it.isString }?.booleanOrNull
}
